File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.*;


class Solution {

    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        int[] answer = new int[enroll.length];
        // 부모,자식 관계를 담기 위한 map 생성
        Map<String, String> recommend = new HashMap<>();
        //수익
        Map<String, Integer> total = new HashMap<>();

        //추천받은 사람(자식) - 추천 한 사람(부모)
        for(int i = 0; i < enroll.length; i++) {
            recommend.put(enroll[i], referral[i]);
        }

        for(int i = 0; i < enroll.length; i++) {
            total.put(enroll[i],0);
        }
        //youn 1200원 들어옴
        //edward를 찾음 -> 1200/10: 120원 young: 1200-120
        //marry를 찾음 -> 120/10 : 12원 edward 120-12
        //민호 찾음 -> 12/10 : 1원 marry 12-1
        for(int i = 0; i<seller.length; i++) {
            int money = amount[i]*100;
            String name = seller[i];
            count(name,money,recommend,total);
        }

        for(int i = 0; i < enroll.length; i++) {
            answer[i] = total.get(enroll[i]);
        }
        return answer;
    }
    public static void count(String name, int money, Map<String, String> recommend, Map<String, Integer>total) {
        //root 만나기 전까지만
        if(name.equals("-")||money<1) {
            return;
        }
        int benefit = money/10;
        int childMoney = money-benefit;
        int updateMoney = total.get(name) + childMoney;
        total.replace(name,updateMoney);
        count(recommend.get(name),benefit,recommend,total);
    }
}

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;
void count(string name, int money, unordered_map<string, string>& recommend, unordered_map<string, int>& total);

vector<int> solution(vector<string> enroll, vector<string> referral, vector<string> seller, vector<int> amount) {
    vector<int> answer(enroll.size());
    //자식-부모 관계
    unordered_map<string, string> recommend;
    //돈 계산
    unordered_map<string, int> total;
    for (int i = 0; i < referral.size(); i++) {
        recommend.insert({ enroll[i], referral[i] });
        total.insert({ enroll[i], 0 });
    }

    for (int i = 0; i < seller.size(); i++) {
        int money = amount[i] * 100;
        string name = seller[i];
        count(name, money, recommend, total);
    }
    for (int i = 0; i < enroll.size(); i++) {
        answer[i] = total[enroll[i]];
    }

    return answer;

}

void count(string name, int money, unordered_map<string, string>& recommend, unordered_map<string, int>& total) {
    if (name == "-" || money < 1)
        return;

    //부모한테 이익 나눠줘야함
    int benefit = money / 10;
    int childMoney = money - benefit;
    total[name] += childMoney;

    count(recommend[name], benefit, recommend, total);
}
